pkg/controller/job/job_controller.go

```diff
@@ -0,0 +1,297 @@
+	// sequence of delays between successive queuings of a service.
+	//
+	// 5ms, 10ms, 20ms, 40ms, 80ms, 160ms, 320ms, 640ms, 1.3s, 2.6s, 5.1s, 10.2s, 20.4s, 41s, 82s
+	maxRetries             = 15
+	revisionsAnnotationKey = "revisions"
+)
+
+var log = logf.Log.WithName("job-controller")
+
+type JobController struct {
+	client           clientset.Interface
+	eventBroadcaster record.EventBroadcaster
+	eventRecorder    record.EventRecorder
+
+	jobLister batchv1listers.JobLister
+	jobSynced cache.InformerSynced
+
+	queue workqueue.RateLimitingInterface
+
+	workerLoopPeriod time.Duration
+}
+
+func NewJobController(jobInformer batchv1informers.JobInformer, client clientset.Interface) *JobController {
+
+	if client != nil && client.CoreV1().RESTClient().GetRateLimiter() != nil {
+		metrics.RegisterMetricAndTrackRateLimiterUsage("job_controller", client.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	v := &JobController{
+		client:           client,
+		queue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "job"),
+		workerLoopPeriod: time.Second,
+	}
+
+	v.jobLister = jobInformer.Lister()
+	v.jobSynced = jobInformer.Informer().HasSynced
+
+	jobInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			v.enqueueJob(obj)
+		},
+		UpdateFunc: func(old, cur interface{}) {
+			v.enqueueJob(cur)
+		},
+	})
+
+	return v
+
+}
+
+func (v *JobController) Start(stopCh <-chan struct{}) error {
+	v.Run(5, stopCh)
+
+	return nil
+}
+
+func (v *JobController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer v.queue.ShutDown()
+
+	log.Info("starting job controller")
+	defer log.Info("shutting down job controller")
+
+	if !controller.WaitForCacheSync("job-controller", stopCh, v.jobSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(v.worker, v.workerLoopPeriod, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (v *JobController) enqueueJob(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+	v.queue.Add(key)
+}
+
+func (v *JobController) worker() {
+	for v.processNextWorkItem() {
+
+	}
+}
+
+func (v *JobController) processNextWorkItem() bool {
+	eKey, quit := v.queue.Get()
+	if quit {
+		return false
+	}
+
+	defer v.queue.Done(eKey)
+
+	err := v.syncJob(eKey.(string))
+	v.handleErr(err, eKey)
+
+	return true
+}
+
+// main function of the reconcile for job
+// job's name is same with the service that created it
+func (v *JobController) syncJob(key string) error {
+	startTime := time.Now()
+	defer func() {
+		log.V(4).Info("Finished syncing job.", "key", key, "duration", time.Since(startTime))
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	job, err := v.jobLister.Jobs(namespace).Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return err
```

Comment:
if err is IsNotFound, normally, means job already been deleted, could just skip this one, no need to return err